plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/SpringReload.java

```diff
@@ -0,0 +1,572 @@
+                // the beanName is not the same as generateBeanName, it should register a new bean definition and remove the old one.
+                if (!beanName.equals(generateBeanName)) {
+                    beanFactory.removeBeanDefinition(beanName);
+                    BeanDefinition newBeanDefinition = new AnnotatedGenericBeanDefinition(clazz);
+                    beanFactory.registerBeanDefinition(generateBeanName, newBeanDefinition);
+                } else {
+                    // remove the attribute, and it will recreate the bean
+                    ((AbstractBeanDefinition) beanDefinition).setBeanClass(clazz);
+                    beanDefinition.removeAttribute("org.springframework.context.annotation.ConfigurationClassPostProcessor.configurationClass");
+                }
+            }
+        }
+    }
+
+    private void destroyBeans() {
+        for (String beanName : new ArrayList<>(recreatedBeans)) {
+            destroyBean(beanName);
+        }
+        // destroy factory bean
+        for (String beanName : beanFactory.getBeanDefinitionNames()) {
+            BeanDefinition beanDefinition = beanFactory.getBeanDefinition(beanName);
+            if (beanDefinition.getFactoryMethodName() != null && beanDefinition.getBeanClassName() != null &&
+                    destroyClasses.contains(beanDefinition.getBeanClassName())) {
+                LOGGER.debug("the bean '{}' will be recreating because the factory class '{}' is changed", beanName, beanDefinition.getBeanClassName());
+                destroyBean(beanName);
+            } else if (beanDefinition.getFactoryMethodName() != null && beanDefinition.getFactoryBeanName() != null &&
+                    destroyBeans.contains(beanDefinition.getFactoryBeanName())) {
+                LOGGER.debug("the bean '{}' will be recreating because the factory bean '{}' is changed", beanName, beanDefinition.getFactoryBeanName());
+                destroyBean(beanName);
+            }
+        }
+    }
+
+    private void destroyBean(String beanName) {
+        // FactoryBean case
+        boolean isFactoryBean = false;
+
+        if (beanName != null && beanName.startsWith("&")) {
+            try {
+                beanFactory.getBeanDefinition(beanName);
+            } catch (NoSuchBeanDefinitionException e) {
+                BeanDefinition mergedBeanDefinition = beanFactory.getMergedBeanDefinition(beanName);
+                if (mergedBeanDefinition instanceof RootBeanDefinition) {
+                    RootBeanDefinition rootBeanDefinition = (RootBeanDefinition) mergedBeanDefinition;
+                    isFactoryBean = BeanFactoryProcessor.isFactoryBean(this.beanFactory, beanName, rootBeanDefinition);
+                }
+                beanName = beanName.substring(1);
+            }
+        }
+        if (destroyBeans.contains(beanName)) {
+            return;
+        }
+        String[] dependentBeans = beanFactory.getDependentBeans(beanName);
```

Comment:
I think we should not care to process a bean's dependent beans since spring framework will take care it for us when a bean definition is destroyed. I unserstand you may want to avoid unnecessary reloading for the dependent beans, but at the same time, the logic below introduces the unnecessary complexicity IMO, especially in 'needDestroy()' method.